app/ide-desktop/lib/dashboard/src/authentication/src/http.tsx

```diff
@@ -0,0 +1,137 @@
+enum HttpMethod {
+  get = "GET",
+  post = "POST",
+  put = "PUT",
+  delete = "DELETE",
+}
+
+// ============
+// === Json ===
+// ============
+
+/** A JavaScript object that can be serialized to a JSON value for an HTTP body. */
+type Json = Record<string, unknown>;
+
+// ====================
+// === HttpResponse ===
+// ====================
+
+/** A wrapper around the built-in JavaScript HTTP {@link Response} type.
+ *
+ * This wrapper provides convenience methods like {@link HttpResponse.model} to deserialize the
+ * response body to a JSON object. */
+export class HttpResponse {
+  /** The wrapped HTTP {@link Response}. */
+  private response: Response;
+
+  /** Creates a new {@link HttpResponse} by wrapping the given {@link Response}. */
+  constructor(response: Response) {
+    this.response = response;
+  }
+
+  /** Returns the status code of the HTTP response. */
+  status = () => this.response.status;
+
+  /** Deserializes the HTTP response body as JSON, to the specified type. */
+  model = <T extends object>(): Promise<T> =>
+    this.response.json().then((json) => json as T);
+}
+
+// ======================
+// === RequestBuilder ===
+// ======================
+
+/** Builder for a new HTTP {@link RequestInitWithUrlAndHeaders}.
+ *
+ * Provides convenience methods for setting the request body, and sending the request using the HTTP
+ * {@link Client} provided when the builder was created. */
+class RequestBuilder {
+  /** HTTP client used to send the HTTP request, once it is created. */
+  private client: Client;
+  /** HTTP request being created. */
+  private request: RequestInitWithUrlAndHeaders;
+
+  /** Creates a new {@link RequestBuilder}. */
+  constructor(client: Client, url: string, config: RequestInit) {
+    const headers = new Headers();
+    this.client = client;
+    this.request = { ...config, url, headers };
+  }
+
+  /** Sets the provided JSON value as the body of the HTTP request, and adjusts the HTTP headers
+   * accordingly. */
+  json = (body: Json) => {
+    this.request.body = JSON.stringify(body);
+    this.request.headers.set("Content-Type", "application/json");
+    return this;
+  };
+
+  /** Finalizes the HTTP request and sends it using the HTTP client. */
+  send = () => this.client.executeRequest(this.request);
+}
+
+// === RequestInitWithUrlAndHeaders ===
+
+/** An HTTP request that is guaranteed to also have the URL it will be send to and an initialized
+ * map of {@link Headers}.
+ *
+ * This is an extension of the built-in {@link RequestInit} interface, which does not provide the
+ * aforementioned guarantees. In fact, it does not provide a `url` at all, and the `headers` field
+ * may even be `undefined` for that interface. */
+interface RequestInitWithUrlAndHeaders extends RequestInit {
+  url: string;
+  headers: Headers;
+}
+
+// ==============
+// === Client ===
+// ==============
+
+/** An HTTP client that can be used to create and send HTTP requests asynchronously. */
+export class Client {
```

Comment:
i have a different implementation of client [here](https://github.com/enso-org/enso/blob/8a3f1836912101e71e7cf870e19dcb97ce90a53d/app/ide-desktop/lib/dashboard/src/authentication/src/http.tsx#L48)

differences:
- no `RequestBuilder`
- no `send`
- different API so it's not possible to accidentally send a body with a `GET` or omit a body from a `POST`

no clue whether it's any better though...

Reply:
(this change would be pretty major so feel free to just close it)

Reply:
This was marked as resolved without providing an answer :) So what was the answer to the questions and what way did you chose?

Reply:
i marked it as resolved because the implementation was changed as suggested. as for why the suggestion was made, it simplifies the logic for this class

Reply:
Yeah, the change was made as suggested.

> no clue whether it's any better though...

Simpler is better, so it was a good suggestion :)